container.go

```diff
@@ -482,6 +482,113 @@ func (c *Client) TopContainer(id string, psArgs string) (TopResult, error) {
+type DetailedMemoryStats struct {
+	TotalPgmajfault         uint64 `json:"total_pgmajfault"`
+	Cache                   uint64 `json:"cache"`
+	MappedFile              uint64 `json:"mapped_file"`
+	TotalInactiveFile       uint64 `json:"total_inactive_file""`
+	Pgpgout                 uint64 `json:"pgpgout"`
+	Rss                     uint64 `json:"rss"`
+	TotalMappedFile         uint64 `json:"total_mapped_file"`
+	Writeback               uint64 `json:"writeback"`
+	Unevictable             uint64 `json:"unevictable"`
+	Pgpgin                  uint64 `json:"pgpgin"`
+	TotalUnevictable        uint64 `json:"total_unevictable"`
+	Pgmajfault              uint64 `json:"pgmajfault"`
+	TotalRss                uint64 `json:"total_rss"`
+	TotalRssHuge            uint64 `json:"total_rss_huge"`
+	TotalWriteback          uint64 `json:"total_writeback"`
+	TotalInactiveAnon       uint64 `json:"total_inactive_anon"`
+	RssHuge                 uint64 `json:"rss_huge"`
+	HierarchicalMemoryLimit uint64 `json:"hierarchical_memory_limit"`
+	TotalPgfault            uint64 `json:"total_pgfault"`
+	TotalActiveFile         uint64 `json:"total_active_file"`
+	TotalPgpgout            uint64 `json:"total_pgpgout"`
+	TotalCache              uint64 `json:"total_cache"`
+	InactiveAnon            uint64 `json:"inactive_anon"`
+	ActiveFile              uint64 `json:"active_file"`
+	Pgfault                 uint64 `json:"pgfault"`
+	InactiveFile            uint64 `json:"inactive_file"`
+	TotalPgpgin             uint64 `json:"pgpgin"`
+}
+
+// CPUStats describes the data related to the container's CPU usage.
+type CPUStats struct {
+	CPUUsage        CPUUsage `json:"cpu_usage"`
+	SystemCPUUsage  uint64   `json:"system_cpu_usage"`
+	PercentageInUse float64
+}
+
+// CPUUsage describes the data related to the container's CPU usage.
+type CPUUsage struct {
+	PercpuUsage       []uint64 `json:"percpu_usage"`
+	UsageInUsermode   uint64   `json:"usage_in_usermode"`
+	TotalUsage        uint64   `json:"total_usage"`
+	UsageInKernelmode uint64   `json:"usage_in_kernelmode"`
+}
+
+// ContainerStats describes the data as it is retrieved from the /stats/ endpoint.
+type ContainerStats struct {
+	Read    string       `json:"read"`
+	Stream  string       `json:"stream,omitempty"`
+	Network NetworkStats `json:"network"`
+	Memory  MemoryStats  `json:"memory_stats"`
+	CPU     CPUStats     `json:"cpu_stats"`
+	EOF     string
+}
+
+// uses a different struct to unmarshal to when we're calling the stats endpoint
+func decodeStats(r io.Reader, stats chan ContainerStats) error {
+	dec := json.NewDecoder(r)
+	for {
+		var m ContainerStats
+		if err := dec.Decode(&m); err != nil {
+			return err
+		}
+		if m.Read != "" {
+			if stats != nil {
+				stats <- m
+			}
+		}
+	}
+	return nil
+}
+
+// StatsContainer gets container stats based on resource usage
+//
+// See http://goo.gl/eY5NRI for more details.
+func (c *Client) StatsContainer(id string, stats chan ContainerStats) error {
```

Comment:
A good practice when receiving a channel is to ensure the intent in this channel, so you could ensure that you're only going to write to the stats chan by declaring it ``stats chan<- ContainerStats``.